Reshard by redistributing samples to new queues

darshanime opened the pull request:
During resharding, try to redistribute the samples amongst the new queues instead of waiting for them all to be sent out, which is prone to tail latency. 

closes https://github.com/prometheus/prometheus/issues/7230

<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

machine424 commented:
Thanks, I'll take a look at this.

bwplotka commented:
(Hello from bug scrub meeting).

Friendly ping @machine424 if you want to review, but it feels that:

* This code is very much a critical path, so we have to be careful.
*  Be mindful of [gained latency improvement vs total latency of remote write](https://github.com/prometheus/prometheus/issues/7230#issuecomment-1917734709)

darshanime commented:
+ @csmarchbanks @cstyan, you might want to review this too as remote maintainers

darshanime commented:
\+  @csmarchbanks @cstyan, you might want to review this too as remote maintainers

cstyan commented:
@darshanime this is on my review list for this week :+1:

cstyan reviewed (commented):
This seems like a good change, but we should do some testing. As Bartek mentioned, we could run into unintended consequences since general throughput is a hot path plus resharding has the potential to cause problems with how long it takes. 

Can we hedge the redistribution? Like, only redistribute if we have more than `# of shards * batch size` samples buffered, or if the flushing takes longer than X seconds (the flush deadline perhaps?)

cstyan commented:
bump @darshanime

darshanime commented:
Thanks for the review, @cstyan 


> Can we hedge the redistribution? Like, only redistribute if we have more than # of shards * batch size samples buffered, or if the flushing takes longer than X seconds (the flush deadline perhaps?)

Currently, we have 2 ways of resharding; triggering soft shutdown, and waiting for all the queues to drain (prone to tail latency amplification), and, if that fails, triggering hard shutdown and dropping samples. This PR adds a 3rd way; redistributing the samples to new shards. The original issue we're trying to address is reducing the impact of resharding; both in terms of time it takes to complete the resharding and dropped samples. With that in mind, imo we should attempt the redistribution first (since it takes "no latency", and causes no dropped samples). We are falling back to the other 2 ways if this fails.

We can hedge the redistribution wrt `# of shards * batch size`, but what is the right number for the threshold? Smaller batched samples are easy+cheap to redistribute, and larger batched samples is where we would reap the most benefit of redistribution. Can we add this hedging later after gaining some experience running this in the wild? I can add a flag to disable redistribution completely, if required.


> This function feels like it's doing a little bit too much IMO. Is there any refactoring we can do so we can reuse stop and start... What do you think of this; rename this function to redistributeSamples

:+1: I've broken the `reshard` function, and created a new function for redistributing the samples.

> In stop we would skip the graceful shutdown attempt and go straight to the unclean shutdown if force is true.

I think this may lead to higher dropped samples if the redistribution fails? Note, when we attempt redistribution and it fails, we haven't spent any time yet, so retaining the original soft -> hard shutdown won't change the current latency characteristics. Can add skipping soft shutdown behind a flag, wdyt?


> the metric for the # of active shards is not being set properly anymore.

Thanks, fixed. I've added it to the `reshard` function so that it can be unit tested easily

cstyan commented on storage/remote/queue_manager.go:
```diff
@@ -1232,20 +1240,95 @@ func (s *shards) stop() {
+	defer s.mtx.Unlock()
+
+	newQueues := make([]*queue, numShards)
+	for i := 0; i < numShards; i++ {
+		newQueues[i] = newQueue(s.qm.cfg.MaxSamplesPerSend, s.qm.cfg.Capacity)
+	}
+
+	successful := s.redistributeSamples(newQueues)
```
One thing I'm thinking about, with the way things are written currently if we are downsharding we could possible often get `successful = false` when attempting to redistribute. 

If we're downsharding and all the existing queues are at full `Capacity`, the newer _lower_ amount of shards won't be able to buffer that many samples. In which case, we either want to know that ahead of time and not attempt to redistribute _or_ have a way to ensure queues that are having samples redistributed to them can send when they're full.

Additionally, we're potentially doubling the amount of memory used by enqueued samples for the duration of the resharding operation.

darshanime replied:
> If we're downsharding and all the existing queues are at full Capacity

If we are near full capacity, `calculateDesiredShards` which takes into consideration `dataPending`, `dataOutRate` etc ought not to reduce the #shards to a very low number? 

Are you proposing that we do some calculation to check if we should attempt to reshard vs falling back to start/stop?

cstyan reviewed (commented):
Are we not also missing a call to `runShard`? We're just creating new queues but passing them to the existing shards?

cstyan commented on storage/remote/queue_manager.go:
```diff
@@ -1232,20 +1240,95 @@ func (s *shards) stop() {
+	hardShutdownCtx, s.hardShutdown = context.WithCancel(context.Background())
+	s.running.Store(int32(numShards))
+	s.done = make(chan struct{})
+	for i := 0; i < numShards; i++ {
+		go s.runShard(hardShutdownCtx, i, newQueues[i])
+	}
```
yeah potentially we want to start these immediately after creating the queues, otherwise we might fail to redistribute when downsharding

bboreham commented:
Hello from the bug-scrub! 

As discussed at length on #7230, I don't agree to add dozens of lines of code to this already very complicated package in order to improve issues from failing networks or infrastructure.  I do agree that we should address #8809.

I see that you updated the code a month ago.

cstyan commented:
I don't want to discourage you @darshanime but I would lean towards Bryan's conclusion as well. 

If there's really a need to deal with resharding costs we can explore other options, like a shorter flush deadline during resharding, but an extended send failure during resharding is likely to reappear after resharding as well. Tail latency for RW requests should still be very low, if it's not that speaks to other issues ( `failing networks or infrastructure` ) IMO.

darshanime commented:
> I would lean towards Bryan's conclusion as well.

Ack; can we have this behind a flag, with the default as false?
